### ISSUE TITLE
Java GeoShape query with envelope does not work in TransportClient

### ISSUE THREAD

**Issue by schulm** [elastic/elasticsearch#34418]
I'm using latest vanilla elasticsearch 6.4.2, this bug is also in version 5.6.8.
A simple geoshapequery with new Envelope returns zero results with the transport client. The same request with the High Level Rest Client works.

A testcase is here: https://github.com/schulm/elasticGeoShapeQueryTestCase

**Comment by colings86** [comment]
@nknize @iverase Could one of you take a look?

**Comment by imotov** [comment]
I can take a look.

**Comment by nknize** [comment]
Thanks @imotov

**Comment by imotov** [comment]
The problem is triggered by an [invalid request](https://github.com/schulm/elasticGeoShapeQueryTestCase/blob/e77f0b4baae92e0c5c8d644f1e05054e4f031621/src/test/java/elastictest/geoShapeNewEnvelope.java#L95) that is sent in both cases (transport and http client):
```
    GeoShapeQueryBuilder getQuery() throws IOException {
        return QueryBuilders.geoShapeQuery(
                "geo", //
                new EnvelopeBuilder(new Coordinate(-21, 44), new Coordinate(-39, 9))).//
                relation(ShapeRelation.WITHIN);
    }
```
The envelope is supposed to be defined by top left and bottom right corners. Instead it is defined by top right and bottom left corners:
```
(-39, 44)                     (-21, 44)


(-39, 9)                      (-21, 9)
```

In case of the HL REST client, we perform an additional check and fix the envelope during json parsing:
https://github.com/elastic/elasticsearch/blob/bd1c513422d63d8f7df1fac6743ba7fd7133fa9e/server/src/main/java/org/elasticsearch/common/geo/GeoShapeType.java#L226-L230

In case of the transport client, since we don't parse json and already have `EnvelopeBuilder` this check is bypassed. To fix the issue, we can either move the coercion logic into  `EnvelopeBuilder` or just add a check that will throw an exception for invalid coordinates in the `EnvelopeBuilder`. @nknize @iverase  what do you think?

@schulm thank you very much for a great reproduction. For now the workaround would be to fix the query and/or move from the transport client to HL REST client since this is a better solution anyway.

**Comment by iverase** [comment]
I disagree with your assessment as I think the provided bounding box is legal. It is totally final that the left longitude is bigger than the right longitude. That only means that the envelope crosses the date line.

You can see the provided envelope as two, divided on the dateline:

```
(-21, 44)   (180, 44)   |  (-180, 44) (-39, 44)
                    
(-21, 9)    (180, 9)    | (-180, 9)  (-39, 9)
```

What the HLRC is doing seems totally wrong, we should not be correcting the provided points. If they make no sense (e.g top latitude lower than bottom latitude) we should throw an exception.

**Comment by imotov** [comment]
@iverase good catch! I keep forgetting about this dateline. You are absolutely right, it looks like the issue is actually in the REST client. 

The following test returns documents `1` in REST and documents `2` and `3` in transport. I agree the correct behaviour should be indeed to return `2` and `3` everywhere.

```
PUT test
{
  "settings": {
    "number_of_shards": 1
  },
  "mappings": {
    "doc": {
      "properties": {
        "geo": {
          "type": "geo_shape"
        }
      }
    }
  }
}

PUT test/doc/1
{
  "geo": {
    "coordinates": [-33.918711, 18.847685],
    "type": "Point"
  }
}

PUT test/doc/2
{
  "geo": {
    "coordinates": [-49.918711, 18.847685],
    "type": "Point"
  }
}


PUT test/doc/3
{
  "geo": {
    "coordinates": [49.918711, 18.847685],
    "type": "Point"
  }
}

GET /test/_search
{
  "query": {
    "geo_shape": {
      "geo": {
        "shape": {
          "type": "envelope",
          "coordinates": [[-21, 44],[-39, 9]]
        },
        "relation": "within"
      }
    }
  }
}
```

So, it looks like a bug in REST client, fixing which can potentially break some applications that were relying on this bug. So, maybe we should fix it only in 6.5 and mention in breaking changes.

**Comment by iverase** [comment]
Sounds like a good plan!
